proposals/NNNN-parameter-packs.md

```diff
@@ -0,0 +1,754 @@
+
+The substituted return type is the tuple type with two elements `(Pair<Int, Bool>, Pair<String, Double>)`, and the returned value is the tuple value with two elements `(Pair(1, true), Pair("hello", 2.0))`.
+
+## Detailed design
+
+**Note:** While this proposal talks about "generic functions", everything also applies to initializers and subscripts nested inside types. With closure expressions, the situation is slightly more limited. Closure expressions support value parameter packs, however since closure expressions do not have polymorphic types in Swift, they're limited to referencing type parameter packs from outer scopes and cannot declare type parameter packs of their own. Also, the value parameter packs of closures cannot have argument labels, because as usual only named declarations have argument labels in Swift.
+
+### Type parameter packs
+
+The generic parameter list of a generic function can contain one or more _type parameter pack declarations_, written as an identifier preceeded by `each`:
+
+```swift
+func variadic<each T, each U>() {}
+```
+
+When referenced from type context, this identifier resolves to a _type parameter pack_. References to type parameter packs can only appear in the following positions:
+
+* The base type of a member type parameter pack, which is again subject to these rules
+* The pattern type of a pack expansion type, where it stands for the corresponding scalar element type
+* The pattern expression of a pack expansion expression, where it stands for the metatype of the corresponding scalar element type and can be used like any other scalar metatype, e.g. to call a static method, call an initializer, or reify the metatype value
+* The subject type of a conformance, superclass, layout, or same-type requirement
+* The constraint type of a same-type requirement
+
+### Pack expansion type
+
+A pack expansion type, written as `repeat P`, has a *pattern type* `P` and a non-empty set of _captured_ type parameter packs.
+
+**Syntactic validity:** Pack expansion types can appear in the following positions:
+
+* The type of a parameter in a function declaration, e.g. `func foo<each T>(values: repeat each T) -> Bool`
+* The type of a parameter in a function type, e.g. `(repeat each T) -> Bool`
+* The type of an unlabeled element in a tuple type, e.g. `(repeat each T)`
+
+The restriction where only unlabeled elements of a tuple type may have a pack expansion type is motivated by ergonomics. If you could write `(t: repeat each T)`, then after a substitution `T := {Int, String}`, the substituted type would be `(t: Int, String)`. This would be strange, because projecting the member `t` would only produce the first element. When an unlabeled element has a pack expansion type, like `(repeat each T)`, then after the above substitution you would get `(Int, String)`. You can still write `0` to project the first element, but this is less surprising to the Swift programmer.
```

Comment:
Hm, I'm not sure I would describe it that way, because labels are not repeated under substitution for functions; you only write an argument label in front of the first pack element. To me, it doesn't feel more intuitive for labels to behave differently between functions and tuples